corehq/change_ownership.py

```diff
@@ -0,0 +1,202 @@
+    batch_end = batch_start + len(case_blocks)
+    percentage_done = round((batch_start / total_count) * 100, 2)
+    print(f'Submitting cases {batch_start}-{batch_end}/{total_count} ({percentage_done}%)')
+    try:
+        submit_cases(case_blocks)
+        write_to_log([cb.case_id for cb in case_blocks])
+        return True
+    except Exception as e:
+        write_to_log(
+            [cb.case_id for cb in case_blocks],
+            is_success=False,
+            message=str(e)
+        )
+        return False
+
+### Task 1
+def move_mobile_workers():
+    print("---MOVING MOBILE WORKER LOCATIONS---")
+    RC_NUM_PROP_NAME = 'rc_number'
+    USER_TYPE_PROP_NAME = 'usertype'
+
+    users = CommCareUser.by_domain(DOMAIN)
+    user_count = len(users)
+    success_count = fail_count = skip_count = 0
+    total_time = 0
+    print(f"Total Users to Process: {user_count}")
+    user_to_save = []
+    for idx, user in enumerate(users):
+        start_time = time.time()
+        percentage_done = round((idx / user_count) * 100, 2)
+        print(f'Processing User {idx}/{user_count} ({percentage_done}%)')
+
+        # First make sure that the user type is rc
+        user_data = user.get_user_data(DOMAIN)
+        if user_data[USER_TYPE_PROP_NAME] != 'rc':
+            write_to_log([user.user_id], is_success=False, message="User Type not RC")
+            skip_count += 1
+            continue
+
+        try:
+            # Get a descendant of user location which has the same rc number
+            loc = SQLLocation.objects.get(
+                domain=DOMAIN,
+                parent__location_id=user.location_id,
+                name=user_data[RC_NUM_PROP_NAME]
+            )
+            # Set this new descendant location as user location
+            user.location_id = loc.location_id
+            user_to_save.append(user)
+            write_to_log([user.user_id])
+            success_count += 1
+        except SQLLocation.DoesNotExist:
+            write_to_log(
+                [user.user_id],
+                is_success=False,
+                message=f'({user_data[RC_NUM_PROP_NAME]}) does not exist as child of location with id ({loc.location_id})'
+            )
+            fail_count += 1
+        finally:
+            # Time logging
+            end_time = time.time()
+            time_diff = end_time - start_time
+            total_time += time_diff
+            print(f'Time to process User #{idx}: {time_diff}s')
+            print(f'Estimate time remaining: {time_diff * (user_count - idx)}s')
+
+        print("Processing Users Complete!")
+        print(
+            f"Success: {success_count}, 
+            Fail: {fail_count}, 
+            Skipped: {skip_count},
+            Total Time: {round(total_time / 60, 2)} minutes"
+        )
+    CommCareUser.bulk_save(user_to_save)
+    print("Saving Users Complete!")
+
+### Task 2 & 3
+def transfer_case_ownership():
+    print("---MOVING CASE OWNERSHIP---")
+    case_ids = (
```

Comment:
I'm curous to get thoughts from others on using ES to get the case IDs. When the script is run, we will be dealing with 11m cases and I feel fetching them all at once in ES could lead to memory problems.

Reply:
Yeah you don't want to try to hold all ids in memory at once. I don't have much experience paginating ES queries, but have done something like this in SQL on CommCareCase. You could use [paginate_query](https://github.com/dimagi/commcare-hq/blob/5a478173565570d926a431eed9814832a3fd2ecd/corehq/sql_db/util.py#L71-L91), with the downside being you'd have to iterate over each partition but that isn't too bad. If helpful, here's an not yet merged [example](https://github.com/dimagi/commcare-hq/pull/34010/files#diff-bdc1e5b639e86fb339dadffb0c55cb680828c5de82cf5898d78de087c4f4e334R124-R125).

Reply:
Thanks for the suggestion, this is very helpful. I've implemented something similar in my approach where I'm storing case IDs in SQLite. The script retrieves the relevant cases using `paginate_query_across_partitioned_databases()`. Once we have the case ids in the DB, we can then easily handle chunking the cases appropriately.